File: Q31.cpp
//Q31.cpp
class Solution {
public:
    void nextPermutation(vector<int>& a) {
        int n=a.size(), i=n-2;
        while(i>=0 && a[i]>=a[i+1]) i--;
        if(i>=0){
            int j=n-1;
            while(a[j]<=a[i]) j--;
            swap(a[i],a[j]);
        }
        reverse(a.begin()+i+1, a.end());
    }
};

File: Q31.py
#Q31.py
class Solution:
    def nextPermutation(self, a):
        n=len(a); i=n-2
        while i>=0 and a[i]>=a[i+1]: i-=1
        if i>=0:
            j=n-1
            while a[j]<=a[i]: j-=1
            a[i],a[j]=a[j],a[i]
        a[i+1:]=reversed(a[i+1:])
